File: projects/basis/sdk-review/js-v2/src/modules/Trading.ts
import { BasisClient } from '../BasisClient';
import ASwapArtifact from '../abis/ASwap.json';
import IERC20Artifact from '../abis/IERC20.json';
import FactoryTokenArtifact from '../abis/FACTORYTOKEN.json';
import { Address, getAddress, getContract } from 'viem';

// Inline ABI for leverage view functions on MAINTOKEN (not in compiled artifact)
const leverageAbi = [
  {"inputs":[{"name":"","type":"address"}],"name":"leverageCount","outputs":[{"name":"","type":"uint256"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"","type":"address"},{"name":"","type":"uint256"}],"name":"leverages","outputs":[{"name":"user","type":"address"},{"name":"token","type":"address"},{"name":"collateralAmount","type":"uint256"},{"name":"liquidatedAmount","type":"uint256"},{"name":"fullAmount","type":"uint256"},{"name":"borrowedAmount","type":"uint256"},{"name":"liquidationTime","type":"uint256"},{"name":"liquidationClaim","type":"uint256"},{"name":"isLiquidated","type":"bool"},{"name":"active","type":"bool"},{"name":"creationTime","type":"uint256"},{"name":"timeOfClosure","type":"uint256"},{"name":"leverage","type":"tuple","components":[{"name":"leverageBuyAmount","type":"uint256"},{"name":"cashedOut","type":"uint256"}]}],"stateMutability":"view","type":"function"},
] as const;

export class TradingModule {
  private client: BasisClient;
  private swapAddress: Address;

  constructor(client: BasisClient, swapAddress: Address) {
    this.client = client;
    this.swapAddress = swapAddress;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  /**
   * Automatically approves the token to be spent by the SWAP contract.
   * Internal helper function.
   */
  private async approveIfNeeded(tokenAddress: Address, amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required for approval.");
    }
    
    const account = this.client.walletClient.account;

    // Check allowance
    const allowance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'allowance',
      args: [account.address, this.swapAddress],
    }) as bigint;

    if (allowance < amount) {
      const { request } = await this.client.publicClient.simulateContract({
        account,
        address: tokenAddress,
        abi: IERC20Artifact.abi,
        functionName: 'approve',
        args: [this.swapAddress, amount],
      });

      const hash = await this.client.walletClient.writeContract(request);
      await this.client.publicClient.waitForTransactionReceipt({ hash });
    }
  }

  /**
   * Buys tokens during the bonding curve phase.
   * Calls buyTokens on SWAP.sol.
   */
  async buyBondingTokens(amount: bigint, minOut: bigint, path: Address[], wrapTokens: boolean) {
    return this.buyTokens(amount, minOut, path, wrapTokens);
  }

  /**
   * Sells tokens during the bonding curve phase.
   * Calls sellTokens on SWAP.sol.
   */
  async sellBondingTokens(amount: bigint, minOut: bigint, path: Address[], swapToETH: boolean) {
    return this.sellTokens(amount, minOut, path, swapToETH);
  }

  /**
   * General buy tokens function.
   */
  async buyTokens(amount: bigint, minOut: bigint, path: Address[], wrapTokens: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const account = this.client.walletClient.account;

    // Approve the input token (path[0])
    if (path.length > 0) {
        await this.approveIfNeeded(path[0], amount);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account,
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'buyTokens',
      args: [amount, minOut, path, wrapTokens],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * General sell tokens function.
   */
  async sellTokens(amount: bigint, minOut: bigint, path: Address[], swapToETH: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const account = this.client.walletClient.account;

    // Approve the input token (path[0]) which is the token being sold
    if (path.length > 0) {
        await this.approveIfNeeded(path[0], amount);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account,
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'sellTokens',
      args: [amount, minOut, path, swapToETH],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Simplified buy: purchases the target token using USDB.
   * Automatically builds the correct swap path.
   */
  async buy(tokenAddress: Address, usdbAmount: bigint, minOut: bigint = 0n, wrapTokens: boolean = false) {
    const path = this.buildBuyPath(tokenAddress);
    return this.buyTokens(usdbAmount, minOut, path, wrapTokens);
  }

  /**
   * Simplified sell: sells a token.
   * For factory tokens, set toUsdb=true to swap all the way to USDB (3-hop),
   * or false to stop at MAINTOKEN (2-hop). Ignored when selling MAINTOKEN.
   */
  async sell(tokenAddress: Address, amount: bigint, toUsdb: boolean = false, minOut: bigint = 0n, swapToETH: boolean = false) {
    const path = this.buildSellPath(tokenAddress, toUsdb);
    return this.sellTokens(amount, minOut, path, swapToETH);
  }

  private buildBuyPath(tokenAddress: Address): Address[] {
    const usdb = getAddress(this.client.usdbAddress);
    const mainToken = getAddress(this.client.mainTokenAddress);
    const target = getAddress(tokenAddress);

    if (target === mainToken) {
      return [usdb, mainToken];
    }
    return [usdb, mainToken, target];
  }

  private buildSellPath(tokenAddress: Address, toUsdb: boolean): Address[] {
    const usdb = getAddress(this.client.usdbAddress);
    const mainToken = getAddress(this.client.mainTokenAddress);
    const target = getAddress(tokenAddress);

    if (target === mainToken) {
      return [mainToken, usdb];
    }
    if (toUsdb) {
      return [target, mainToken, usdb];
    }
    return [target, mainToken];
  }

  /**
   * Leveraged buy: purchases tokens with leverage (creates a loan position).
   */
  async leverageBuy(amount: bigint, minOut: bigint, path: Address[], numberOfDays: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const account = this.client.walletClient.account;

    // Approve USDB for the swap contract
    if (path.length > 0) {
      await this.approveIfNeeded(path[0], amount);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account,
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'leverageBuy',
      args: [amount, minOut, path, numberOfDays],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Partially sells collateral from a loan/leverage position.
   * percentage must be divisible by 10 (10-100).
   */
  async partialLoanSell(loanId: bigint, percentage: bigint, isLeverage: boolean, minOut: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'partialLoanSell',
      args: [loanId, percentage, isLeverage, minOut],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Sells a percentage of the user's token balance.
   * percentage: 1-100
   */
  async sellPercentage(tokenAddress: Address, percentage: number, toUsdb: boolean = false, minOut: bigint = 0n, swapToETH: boolean = false) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }
    if (percentage < 1 || percentage > 100) {
      throw new Error("Percentage must be between 1 and 100.");
    }

    const account = this.client.walletClient.account;

    // Read current balance
    const balance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'balanceOf',
      args: [account.address],
    }) as bigint;

    if (balance === 0n) {
      throw new Error("Token balance is zero.");
    }

    const sellAmount = (balance * BigInt(percentage)) / 100n;
    return this.sell(tokenAddress, sellAmount, toUsdb, minOut, swapToETH);
  }

  /**
   * Gets the leverage position count for a user from MAINTOKEN.
   */
  async getLeverageCount(user: Address) {
    const count = await this.client.publicClient.readContract({
      address: this.client.mainTokenAddress,
      abi: leverageAbi,
      functionName: 'leverageCount',
      args: [user],
    }) as bigint;
    return count;
  }

  /**
   * Gets a specific leverage position from MAINTOKEN.
   */
  async getLeveragePosition(user: Address, loanId: bigint) {
    return this.client.publicClient.readContract({
      address: this.client.mainTokenAddress,
      abi: leverageAbi,
      functionName: 'leverages',
      args: [user, loanId],
    });
  }

  /**
   * Fetches the token price from the token's contract.
   */
  async getTokenPrice(tokenAddress: Address) {
    const price = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'getTokenPrice',
    }) as bigint;
    
    return price.toString();
  }

  /**
   * Fetches the USD price of the token from the token's contract.
   */
  async getUSDPrice(tokenAddress: Address) {
    const price = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'getUSDPrice',
    }) as bigint;

    return price.toString();
  }

  /**
   * Converts a market token position to native tokens via the swap contract.
   * Auto-approves the input token.
   */
  async convertToNative(marketToken: Address, inputToken: Address, inputAmount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Auto-approve inputToken to swap contract
    await this.approveIfNeeded(inputToken, inputAmount);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'convertToNative',
      args: [marketToken, inputToken, inputAmount],
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Returns the expected output amounts for a given input amount and swap path.
   */
  async getAmountsOut(amount: bigint, path: Address[]): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.swapAddress,
      abi: ASwapArtifact.abi,
      functionName: 'getAmountsOut',
      args: [amount, path],
    }) as Promise<bigint>;
  }
}

File: projects/basis/sdk-review/py-v2/basis/modules/trading.py
import logging
from web3 import Web3
from .factory import load_abi

logger = logging.getLogger(__name__)


class TradingModule:
    def __init__(self, client, swap_address: str):
        self.client = client
        self.swap_address = Web3.to_checksum_address(swap_address)
        self.swap_abi = load_abi('ASwap.json')
        self.erc20_abi = load_abi('IERC20.json')
        self.token_abi = load_abi('FACTORYTOKEN.json')
        self.maintoken_abi = load_abi('MAINTOKEN.json')
        self.contract = self.client.web3.eth.contract(address=self.swap_address, abi=self.swap_abi)

    def _sync_tx(self, tx_hash: str):
        """Sync tx to backend. Non-fatal on failure."""
        try:
            if not tx_hash.startswith("0x"):
                tx_hash = "0x" + tx_hash
            self.client.api.sync_transaction(tx_hash)
        except Exception as e:
            logger.warning("Sync warning: %s", e)

    def _approve_if_needed(self, token_address: str, amount: int):
        if not self.client.account:
            raise ValueError("Wallet account is required for approval.")

        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.erc20_abi)
        
        allowance = token_contract.functions.allowance(
            self.client.account.address, self.swap_address
        ).call()

        if allowance < amount:
            func = token_contract.functions.approve(self.swap_address, amount)
            tx = func.build_transaction({
                'from': self.client.account.address,
                'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            })
            signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
            tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
            self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

    def _build_and_send_tx(self, function_call):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        
        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        
        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _build_buy_path(self, token_address: str) -> list[str]:
        usdb = self.client.usdb_address
        main_token = self.client.main_token_address
        target = Web3.to_checksum_address(token_address)

        if target == main_token:
            return [usdb, main_token]
        return [usdb, main_token, target]

    def _build_sell_path(self, token_address: str, to_usdb: bool) -> list[str]:
        usdb = self.client.usdb_address
        main_token = self.client.main_token_address
        target = Web3.to_checksum_address(token_address)

        if target == main_token:
            return [main_token, usdb]
        if to_usdb:
            return [target, main_token, usdb]
        return [target, main_token]

    def buy(self, token_address: str, usdb_amount: int, min_out: int = 0, wrap_tokens: bool = False):
        """Simplified buy: purchases the target token using USDB. Builds path automatically."""
        path = self._build_buy_path(token_address)
        return self.buy_tokens(usdb_amount, min_out, path, wrap_tokens)

    def sell(self, token_address: str, amount: int, to_usdb: bool = False, min_out: int = 0, swap_to_eth: bool = False):
        """Simplified sell: sells a token. Set to_usdb=True to swap factory tokens all the way to USDB."""
        path = self._build_sell_path(token_address, to_usdb)
        return self.sell_tokens(amount, min_out, path, swap_to_eth)

    def buy_bonding_tokens(self, amount: int, min_out: int, path: list[str], wrap_tokens: bool):
        return self.buy_tokens(amount, min_out, path, wrap_tokens)

    def sell_bonding_tokens(self, amount: int, min_out: int, path: list[str], swap_to_eth: bool):
        return self.sell_tokens(amount, min_out, path, swap_to_eth)

    def buy_tokens(self, amount: int, min_out: int, path: list[str], wrap_tokens: bool):
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        if checksum_path:
            self._approve_if_needed(checksum_path[0], amount)
            
        func = self.contract.functions.buyTokens(amount, min_out, checksum_path, wrap_tokens)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def sell_tokens(self, amount: int, min_out: int, path: list[str], swap_to_eth: bool):
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        if checksum_path:
            self._approve_if_needed(checksum_path[0], amount)
            
        func = self.contract.functions.sellTokens(amount, min_out, checksum_path, swap_to_eth)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def get_token_price(self, token_address: str):
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        price = token_contract.functions.getTokenPrice().call()
        return str(price)

    def get_usd_price(self, token_address: str):
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        price = token_contract.functions.getUSDPrice().call()
        return str(price)

    def leverage_buy(self, amount: int, min_out: int, path: list[str], number_of_days: int):
        """Leveraged buy: purchases tokens with leverage (creates a loan position)."""
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        if checksum_path:
            self._approve_if_needed(checksum_path[0], amount)
        func = self.contract.functions.leverageBuy(amount, min_out, checksum_path, number_of_days)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def partial_loan_sell(self, loan_id: int, percentage: int, is_leverage: bool, min_out: int = 0):
        """Partially sells collateral from a loan/leverage position. percentage must be divisible by 10 (10-100)."""
        func = self.contract.functions.partialLoanSell(loan_id, percentage, is_leverage, min_out)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def sell_percentage(self, token_address: str, percentage: int, to_usdb: bool = False, min_out: int = 0, swap_to_eth: bool = False):
        """Sells a percentage (1-100) of the user's token balance."""
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        if percentage < 1 or percentage > 100:
            raise ValueError("Percentage must be between 1 and 100.")

        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.erc20_abi)
        balance = token_contract.functions.balanceOf(self.client.account.address).call()

        if balance == 0:
            raise ValueError("Token balance is zero.")

        sell_amount = (balance * percentage) // 100
        return self.sell(token_address, sell_amount, to_usdb, min_out, swap_to_eth)

    _LEVERAGE_ABI = [
        {"inputs":[{"name":"","type":"address"}],"name":"leverageCount","outputs":[{"name":"","type":"uint256"}],"stateMutability":"view","type":"function"},
        {"inputs":[{"name":"","type":"address"},{"name":"","type":"uint256"}],"name":"leverages","outputs":[{"name":"user","type":"address"},{"name":"token","type":"address"},{"name":"collateralAmount","type":"uint256"},{"name":"liquidatedAmount","type":"uint256"},{"name":"fullAmount","type":"uint256"},{"name":"borrowedAmount","type":"uint256"},{"name":"liquidationTime","type":"uint256"},{"name":"liquidationClaim","type":"uint256"},{"name":"isLiquidated","type":"bool"},{"name":"active","type":"bool"},{"name":"creationTime","type":"uint256"},{"name":"timeOfClosure","type":"uint256"},{"name":"leverage","type":"tuple","components":[{"name":"leverageBuyAmount","type":"uint256"},{"name":"cashedOut","type":"uint256"}]}],"stateMutability":"view","type":"function"},
    ]

    def get_leverage_count(self, user: str) -> int:
        """Gets the leverage position count for a user from MAINTOKEN."""
        checksum_user = Web3.to_checksum_address(user)
        contract = self.client.web3.eth.contract(
            address=self.client.main_token_address, abi=self._LEVERAGE_ABI
        )
        return contract.functions.leverageCount(checksum_user).call()

    def get_leverage_position(self, user: str, loan_id: int):
        """Gets a specific leverage position from MAINTOKEN."""
        checksum_user = Web3.to_checksum_address(user)
        contract = self.client.web3.eth.contract(
            address=self.client.main_token_address, abi=self._LEVERAGE_ABI
        )
        return contract.functions.leverages(checksum_user, loan_id).call()

    def convert_to_native(self, market_token: str, input_token: str, input_amount: int):
        """Converts a market token position to native tokens via the swap contract. Auto-approves the input token."""
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_input = Web3.to_checksum_address(input_token)
        self._approve_if_needed(checksum_input, input_amount)
        func = self.contract.functions.convertToNative(checksum_market, checksum_input, input_amount)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def get_amounts_out(self, amount: int, path: list[str]) -> int:
        """Returns the estimated output amounts for a given input amount and swap path."""
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        return self.contract.functions.getAmountsOut(amount, checksum_path).call()
